Backend/Remora.Discord.Voice/DiscordVoiceClient.cs

```diff
@@ -0,0 +1,1258 @@
+                        opusBuffer.Memory[0..encodeResult.Entity],
+                        read,
+                        ct
+                    ).ConfigureAwait(false);
+
+                    if (!sendFrameResult.IsSuccess)
+                    {
+                        return sendFrameResult;
+                    }
+                }
+
+                var sendSilence = SendSilenceFrames(sampleSize, ct);
+                if (!sendSilence.IsSuccess)
+                {
+                    return sendSilence;
+                }
+
+                return Result.FromSuccess();
+            }
+            catch (Exception ex)
+            {
+                return ex;
+            }
+            finally
+            {
+                inputBuffer.Dispose();
+                opusBuffer.Dispose();
+
+                EnqueueCommand
+                (
+                    new VoiceSpeakingCommand
+                    (
+                        SpeakingFlags.None,
+                        0,
+                        ssrc
+                    )
+                );
+
+                if (needsRelease)
+                {
+                    _transmitSemaphore.Release();
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
+        {
+            if (this.ConnectionStatus is not GatewayConnectionStatus.Offline)
+            {
+                await StopAsync().ConfigureAwait(false);
+            }
+
+            _transmitSemaphore.Dispose();
+            _silenceFrameBuffer.Dispose();
+        }
+
+        /// <summary>
+        /// Restores the client to a near pre-startup state, intended for when the client is stopping or has encountered a fatal error.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private async Task CleanupAsync()
+        {
+            if (!_disconnectRequestedSource.IsCancellationRequested)
+            {
+                _disconnectRequestedSource.Cancel();
+            }
+
+            await Task.WhenAll(_sendTask, _receiveTask, _runLoopTask).ConfigureAwait(false);
+            _sendTask.Dispose();
+            _receiveTask.Dispose();
+            _runLoopTask.Dispose();
+
+            if (_transportService.IsConnected)
+            {
+                await _transportService.DisconnectAsync(false).ConfigureAwait(false);
+            }
+
+            if (_dataService.IsConnected)
+            {
+                _dataService.Disconnect();
+            }
+
+            _disconnectRequestedSource.Dispose();
+            _disconnectRequestedSource = new CancellationTokenSource();
+
+            _payloadsToSend.Clear();
+            _receivedPayloads.Clear();
+            _connectionRequestParameters = null;
+            _gatewayConnectionDetails = null;
+            _voiceServerConnectionDetails = null;
+
+            this.ConnectionStatus = GatewayConnectionStatus.Offline;
```

Comment:
So, I was gonna comment on previous lines about setting the connection status, but I see it here too.

Is there a reason for setting to offline (In ConnectAsync I believe)?

Reply:
Actually I may have seen `Disconnected`, correct me if the above comment is wrong.

Reply:
I'm not sure I entirely follow sorry. The `ConnectionStatus` property is used for internal connection control; if we're offline we need to open a new connection and if we're disconnected we need to resume.
The property isn't updated at all in `ConnectAsync` and the instance you've highlighted is in the `CleanupAsync` method, which we perform on shutdown.